src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -384,6 +384,31 @@ case object EXTCODECOPY extends OpCode(0x3c, 4, 0, G_extcode) {
   }
 }
 
+case object BLOCKHASH extends OpCode(0x40, 1, 1, G_blockhash) with ConstGas {
+  protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
+    val (blockNumber, stack1) = state.stack.pop
+
+    if (state.context.env.blockHeader.number - blockNumber > 256 || blockNumber > state.context.env.blockHeader.number) {
+      val stack2 = stack1.push(DataWord(0))
+      state.withStack(stack2).step()
+    } else {
+      val result = state.world.getBlockHash(blockNumber).map(DataWord(_)).getOrElse(DataWord(0))
+      val stack2 = stack1.push(result)
+      state.withStack(stack2).step()
+    }
```

Comment:
This could be slightly improved so that we don't repeat the stack and state updates, in psuedo:

```
val hash = if (conditionsMet) getBlockhash(number).getOrElse(DataWord.Zero) else DataWord.Zero
val stack2 = stack1.push(result)
state.withStack(stack2).step()
```